src/ipc/ipc4/helper.c

```diff
@@ -880,11 +881,43 @@ int ipc4_find_dma_config(struct ipc_config_dai *dai, uint8_t *data_buffer, uint3
 	if (*dma_config_id != GTW_DMA_CONFIG_ID)
 		return IPC4_INVALID_REQUEST;
 
-	dai->host_dma_config = GET_IPC_DMA_CONFIG(data_buffer, size);
+	dai->host_dma_config[0] = GET_IPC_DMA_CONFIG(data_buffer, size);
 #endif
 	return IPC4_SUCCESS;
 }
 
+int ipc4_find_dma_config_multiple(struct ipc_config_dai *dai, uint8_t *data_buffer,
+				  uint32_t size, uint32_t device_id)
+{
+#if defined(CONFIG_ACE_VERSION_2_0)
+	uint32_t end_addr = (uint32_t)data_buffer + size;
+	struct ipc_dma_config *dma_cfg;
+	struct sof_tlv *tlvs;
+	int dma_cfg_count = 0;
+
+	for (tlvs = (struct sof_tlv *)data_buffer; (uint32_t)tlvs < end_addr;
+	     tlvs = tlv_next(tlvs)) {
+		dma_cfg = (struct ipc_dma_config *)tlv_value_ptr_get(
+			   tlvs, GTW_DMA_CONFIG_ID);
+
+		if (!dma_cfg)
+			continue;
+
+		for (uint32_t i = 0; i < dma_cfg->channel_map.device_count; i++) {
+			if (dma_cfg->channel_map.map[i].device_address == device_id) {
+				dai->host_dma_config[dma_cfg_count] = dma_cfg;
```

Comment:
this is a somewhat strange counter: `dma_cfg_count` how many valid (with a non-NULL `dma_cfg` pointer) didn't match the requested device and then assigns to an array with that index... The connection isn't obvious. Could you clarify?

Reply:
that is absolutely not what we discussed with @mmaka1. device_address was supposed to be the PCMsCMy FIFO address.

Reply:
Yes, device_address should be PCMsCMy, but Windows drv always set this value and I need to deal with it.

Reply:
I am still lost by all this.

what exactly is this 'device_id'

a) the link_id as I can read in the November 3 comment
b) the address of the PCMsCMy register
c) the PDI index written in PCMsCMy?

Also we all agree that the firmware does not program the PCMsCMy register, so what exactly does the firmware do with the device_id.

How many engineers does it take to define what a 'device' is?


Two additional questions: what happens if we do

One additional question: the 'device_id' aka PDI ID is specific to a link. The DMA configuration is not link specific, we can use whatever stream_tag we want.

So how do we know which PDI